### PR TITLE
perf: replace nullish coalescing operator and optional chaining

### PR DESCRIPTION
**Pull request by hchlq**
[[中文版模板 / Chinese template](https://github.com/alibaba/hooks/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a Performance optimization.

- [ ] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [x] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |           |
| 🇨🇳 Chinese |           |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

### CODE REVIEW

**Comment by hchlq** [comment]
nullish coalescing operator (??) and optional chaining (?.) will transpile it to something much more verbose which increased code size.

Although the code looks more concise when we write the code, it is more redundant after actual packaging, which will increase the volume of the package and may affect the loading speed of ahooks

about 2KB less
image

**Comment by hchlq** [comment]
such as 
```js
const wait = options?.wait ?? 1000;
```
will compile into
```js
var wait = (_a = options === null || options === void 0 ? void 0 : options.wait) !== null && _a !== void 0 ? _a : 1000;
```

If the optional operator is not used, such as
```js
const wait = (options && options.wait != undefined) ? options.wait : 1000
```
will compile into, the code is more concise.
```js
var wait = options && options.wait != undefined ? options.wait : 1000;
```

File: .eslintrc.js [hunk 1]
```diff
@@ -12,6 +12,7 @@ module.exports = {
     '@typescript-eslint/ban-ts-ignore': 'off',
     '@typescript-eslint/no-object-literal-type-assertion': 'off',
     '@typescript-eslint/no-parameter-properties': 'off',
+    '@typescript-eslint/no-unused-expressions': 'off',
```
Comment (hchlq) [inline .eslintrc.js:15]: such 
```js
// such as this expression will error in eslint, so emit this error
onEnter && onEnter()
```

**Comment by awmleer** [comment]
Thanks for contributing! 😅 However, I'm going to close this PR, considering:

- The way `tsc` compiles nullish coalescing and optional chaining are guaranteed to strictly follow the ES grammar standard. While the simple `foo != undefined` syntax may introduce some bugs in edge cases.
- For most of the projects 2kb (maybe 1kb or less after gzip) is not really a big deal. But runtime bugs are.
- In the future we may not need to do the transpiling of nullish coalescing and optional chaining. Currently there are plenty of browsers that natively support these syntax.

**Pull request closed without merging**